components/encryption/src/log_file.rs

```diff
@@ -0,0 +1,387 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use byteorder::{BigEndian, ByteOrder};
+use kvproto::encryptionpb::{EncryptedContent, FileDictionary, FileInfo};
+use protobuf::Message;
+use rand::{thread_rng, RngCore};
+
+use crate::encrypted_file::{Header, Version, TMP_FILE_SUFFIX};
+use crate::master_key::{Backend, PlaintextBackend};
+use crate::Result;
+
+use std::fs::{rename, File, OpenOptions};
+use std::io::BufRead;
+use std::io::{Read, Write};
+use std::path::{Path, PathBuf};
+
+#[derive(Debug)]
+enum LogRecord {
+    INSERT(FileInfo),
+    REMOVE,
+}
+
+/// LogFile is used to store log style file dictionary.
+///
+/// Layout:
+/// ```text
+/// Encrypted File Header + FileDictionary Content | Record Header + Record Content | etc
+/// ```
+#[derive(Debug)]
+pub struct LogFile {
+    base: PathBuf,
+    name: String,
+    // Avoid reopening the file every time when append.
+    append_file: Option<File>,
+    // Internal file dictionary to avoid recovery before rewrite and to check if compaction
+    // is needed.
+    file_dict: FileDictionary,
```

Comment:
a little bit sad we have another copy of file_dict, which increase the chance of inconsistency if we have bug in the code. but let it be.